src/electron/routing_service.ts

```diff
@@ -46,136 +49,115 @@ enum RoutingServiceStatusCode {
   UNSUPPORTED_ROUTING_TABLE = 2
 }
 
-// Define the error type thrown by the net module.
-interface NetError extends Error {
-  code?: string|number;
-  errno?: string;
-  syscall?: string;
-  address?: string;
-}
-
-// Abstracts IPC with OutlineService in order to configure routing.
-export class RoutingService {
-  // TODO: make this a constructor argument
-  private ipcConnection: net.Socket|null = null;
-
-  // Asks OutlineService to configure all traffic, except that bound for the proxy server,
-  // to route via routerIp.
-  configureRouting(
-      routerIp: string, proxyIp: string, onStatusChange: (status: ConnectionStatus) => void,
-      isAutoConnect = false): Promise<string> {
-    return this.sendRequest(
-        {
-          action: RoutingServiceAction.CONFIGURE_ROUTING,
-          parameters: {proxyIp, routerIp, isAutoConnect}
-        },
-        true, onStatusChange);
-  }
-
-  // Restores the default system routes.
-  resetRouting(): Promise<string> {
-    try {
-      if (this.ipcConnection) {
-        this.ipcConnection.removeAllListeners();
-      }
-    } catch (e) {
-      // Ignore, the service may have disconnected the pipe.
-    }
-    return this.sendRequest({action: RoutingServiceAction.RESET_ROUTING, parameters: {}});
-  }
+// Communicates with the Outline routing daemon via a Unix socket.
+//
+// A minimal life-cycle is supported:
+//  - CONFIGURE_ROUTING is *always* the first message sent on the pipe.
+//  - The only subsequent supported operation is RESET_ROUTING.
+//  - In the meantime, the client may receive zero or more STATUS_CHANGED events.
+//
+// That's it! This helps us connect to the service for *as short a time as possible* which is
+// important when trying to implement a Promise-like interface over what is essentially a pipe *and*
+// on Windows where only one client may be connected to the service at any given time.
+//
+// To test:
+//  - Linux: systemctl start|stop outline_proxy_controller.service
+//  - Windows: net start|stop OutlineService
+export class RoutingDaemon {
+  // Fulfills once a connection is established with the routing daemon *and* it has successfully
+  // configured the system's routing table.
+  static async create(proxyAddress: string, isAutoConnect: boolean, retry = true) {
```

Comment:
The static `create` methods in `RoutingDaemon` and `ConnectionManager`, along with their constructors, create the object (state) and also perform actions (side-effect). Could we decouple the object creation from doing the actual work?

Reply:
I played around with this and I think I like the `start` methods better - well, it's a tradeoff and generally I prefer immutable objects but this is more in line with the rest of the code (I think I might have settled on that approach because of `net.createConnection` in the RoutingDaemon class, but it's fine). Hope you like it!